modules/standard/GPU.chpl

```diff
@@ -389,4 +389,155 @@ module GPU
+    }
+
+
+    proc chplTypeToCTypeName(type t) param {
+      select t {
+        when int(8)   do return "int8_t";
+        when int(16)  do return "int16_t";
+        when int(32)  do return "int32_t";
+        when int(64)  do return "int64_t";
+        when uint(8)  do return "uint8_t";
+        when uint(16) do return "uint16_t";
+        when uint(32) do return "uint32_t";
+        when uint(64) do return "uint64_t";
+        when real(32) do return "float";
+        when real(64) do return "double";
+      }
+      return "unknown";
+    }
+
+    proc getExternFuncName(param op: string, type t) param: string {
+      return "chpl_gpu_"+op+"_reduce_"+chplTypeToCTypeName(t);
+    }
+
+    inline proc subReduceVal(param op, ref accum: ?t, val: t) {
+      select op {
+        when "sum" do accum += val;
+        when "min" do accum = min(accum, val);
+        when "max" do accum = max(accum, val);
+      }
+    }
+
+    inline proc subReduceValIdx(param op, const baseOffset, ref accum: ?t,
+                                val: t) {
+      select op {
+        when "minloc" do
+          if accum[1] > val[1] then accum = (val[0]+baseOffset, val[1]);
+        when "maxloc" do
+          if accum[1] < val[1] then accum = (val[0]+baseOffset, val[1]);
+      }
+    }
+
+    iter offsetsThatCanFitIn32Bits(size: int) {
+      // Engin: I've tried to get max(int(32)) to work as this bug is about CUB
+      // using `int` as the size in the interface. However, getting close to
+      // max(int(32)) also triggers the bug. So, I am choosing this as a
+      // round/safe value for the time being.
+      param chunkSize = 2_000_000_000;
+
+      use Math only divCeil;
+      const numChunks = divCeil(size, chunkSize);
+      const standardChunkSize = divCeil(size, numChunks);
+
+      if gpuDebugReduce then
+        writeln("Will use ", numChunks, " chunks of size ", standardChunkSize);
+
+      foreach chunk in 0..<numChunks {
+        const start = chunk*standardChunkSize;
+        const curChunkSize = if start+standardChunkSize <= size
+                               then standardChunkSize
+                               else size-start;
+        if gpuDebugReduce then
+          writef("Chunk %i: (start=%i, curChunkSize=%i) ", chunk, start,
+                 curChunkSize);
+
+        yield (start, curChunkSize);
+      }
+    }
+
+    use CTypes;
+
+    param externFunc = getExternFuncName(op, t);
+
+    if op == "sum" || op == "min" || op == "max" {
+      var val: t;
+      if op == "min" then
+        val = max(t);
+      else if op == "max" then
+        val = min(t);
+
+      extern externFunc proc reduce_fn(data, size, ref val);
+
+      const basePtr = c_ptrToConst(A);
+      for (offset,size) in offsetsThatCanFitIn32Bits(A.size) {
+        var curVal: t;
+        reduce_fn(basePtr+offset, size, curVal);
+        if gpuDebugReduce then
+          writef(" (curVal=%i)\n", curVal);
+        subReduceVal(op, val, curVal);
```

Comment:
1675b9e819 was a really pleasing refactor for this. Note that I had to go deeper so that the runtime interface is unified across reduction types at least at the top level. But that was also a good thing by itself.